Python/Product/PythonTools/PythonTools/Environments/EnvironmentSwitcherManager.cs

```diff
@@ -0,0 +1,277 @@
+        private readonly IInterpreterOptionsService _optionsService;
+        private readonly IInterpreterRegistryService _registryService;
+        private readonly IVsMonitorSelection _monitorSelection;
+        private readonly IVsShell _shell;
+        private readonly IVsFolderWorkspaceService _workspaceService;
+        private uint _selectionEventsCookie;
+        private IVsHierarchy _previousHier;
+        private IEnvironmentSwitcherContext _context;
+        private bool _isInitialized;
+        private bool _isPythonWorkspace;
+
+        public event EventHandler EnvironmentsChanged;
+
+        public EnvironmentSwitcherManager(IServiceProvider serviceProvider) {
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            _optionsService = serviceProvider.GetComponentModel().GetService<IInterpreterOptionsService>();
+            _registryService = serviceProvider.GetComponentModel().GetService<IInterpreterRegistryService>();
+            _monitorSelection = serviceProvider.GetService(typeof(IVsMonitorSelection)) as IVsMonitorSelection;
+            _shell = serviceProvider.GetService(typeof(SVsShell)) as IVsShell;
+            _workspaceService = serviceProvider.GetComponentModel().GetService<IVsFolderWorkspaceService>();
+            AllFactories = Enumerable.Empty<IPythonInterpreterFactory>();
+        }
+
+        public bool IsClosing { get; set; }
+
+        /// <summary>
+        /// Returns whether a Python project, workspace or document is currently
+        /// selected / active. Used to control the visibility of the status bar
+        /// switcher.
+        /// </summary>
+        public bool IsInPythonMode => Context != null;
+
+        /// <summary>
+        /// Interpreter factory that is currently applicable to the current
+        /// project, workspace, or loose document. This may be null even when
+        /// <see cref="IsInPythonMode"/> is <c>true</c>, for example when no
+        /// environment is found on the machine.
+        /// </summary>
+        public IPythonInterpreterFactory CurrentFactory { get; private set; }
+
+        /// <summary>
+        /// List of factories to present to the user, based on the current
+        /// project, workspace or loose document.
+        /// </summary>
+        /// <remarks>
+        /// For a project, this is the list of environments referenced by the
+        /// project. For a loose document or workspace, this is the list of all
+        /// environments found on the machine. 
+        /// </remarks>
+        public IEnumerable<IPythonInterpreterFactory> AllFactories { get; private set; }
+
+        public IEnvironmentSwitcherContext Context {
+            get => _context;
+            private set {
+                if (_context != null) {
+                    _context.EnvironmentsChanged -= OnInterpretersChanged;
+                    _context.Dispose();
```

Comment:
That is a dangerous approach to dispose object in property setter. It is unexpected for property callers and is responsibility of the code that manages life cycle of the contexts